crates/blockifier/src/execution/native_syscall_handler.rs

```diff
@@ -80,31 +212,96 @@ impl<'state> StarkNetSyscallHandler for NativeSyscallHandler<'state> {
+        keys: &[Felt],
+        data: &[Felt],
         _remaining_gas: &mut u128,
     ) -> cairo_native::starknet::SyscallResult<()> {
-        todo!("Native syscall handler - emit_event")
+        let order = self.execution_context.n_emitted_events;
+
+        self.events.push(OrderedEvent {
+            order,
+            event: EventContent {
+                keys: keys
+                    .to_vec()
+                    .iter()
+                    .map(|felt| EventKey(felt_to_starkfelt(*felt)))
+                    .collect::<Vec<EventKey>>(),
+                data: EventData(
+                    data.to_vec().iter().map(|felt| felt_to_starkfelt(*felt)).collect(),
+                ),
+            },
+        });
+
+        self.execution_context.n_emitted_events += 1;
+
+        Ok(())
     }
 
     fn send_message_to_l1(
         &mut self,
-        _to_address: Felt,
-        _payload: &[Felt],
+        to_address: Felt,
+        payload: &[Felt],
         _remaining_gas: &mut u128,
     ) -> cairo_native::starknet::SyscallResult<()> {
-        todo!("Native syscall handler - send_message_to_l1")
+        let order = self.execution_context.n_sent_messages_to_l1;
+
+        self.l2_to_l1_messages.push(OrderedL2ToL1Message {
+            order,
+            message: MessageToL1 {
+                // todo: handle error properly
+                to_address: EthAddress::try_from(felt_to_starkfelt(to_address)).unwrap(),
+                payload: L2ToL1Payload(
+                    payload.iter().map(|felt| felt_to_starkfelt(*felt)).collect(),
+                ),
+            },
+        });
+
+        self.execution_context.n_sent_messages_to_l1 += 1;
+
+        Ok(())
     }
 
-    fn keccak(&mut self, _input: &[u64], _remaining_gas: &mut u128) -> cairo_native::starknet::SyscallResult<cairo_native::starknet::U256> {
-        todo!("Native syscall handler - keccak")
+    fn keccak(
+        &mut self,
+        input: &[u64],
+        _remaining_gas: &mut u128,
+    ) -> cairo_native::starknet::SyscallResult<cairo_native::starknet::U256> {
+        let input_len = input.len();
+
+        const KECCAK_FULL_RATE_IN_WORDS: usize = 17;
+        let (_, remainder) = num_integer::div_rem(input_len, KECCAK_FULL_RATE_IN_WORDS);
+
+        if remainder != 0 {
+            return Err(vec![Felt::from_hex(INVALID_INPUT_LENGTH_ERROR).unwrap()]);
+        }
+
+        let input_chunks = input.chunks_exact(KECCAK_FULL_RATE_IN_WORDS);
+        let mut keccak_state = [0u64; 25];
+
+        for chunk in input_chunks {
+            for (i, val) in chunk.iter().enumerate() {
+                keccak_state[i] ^= val;
+            }
+            keccak::f1600(&mut keccak_state)
+        }
+
+        let hash: Vec<Vec<u8>> =
+            [keccak_state[0], keccak_state[1], keccak_state[2], keccak_state[3]]
+                .iter()
+                .map(|e| e.to_le_bytes().to_vec())
+                .collect();
+
+        let hash = hash.concat();
```

Comment:
The rationale behind the suggestion is that I think it is easier to read, but feel free to ignore it
```suggestion
        let hash = [
            keccak_state[0].to_le_bytes(),
            keccak_state[1].to_le_bytes(),
            keccak_state[2].to_le_bytes(),
            keccak_state[3].to_le_bytes(),
        ]
        .concat();
```

Reply:
I thought about that, but decided to stick with DRY :D